File: src/main/java/com/tkr/tools/TkrCtl.java
package com.tkr.tools;

import com.tkr.desk.ProRataAllocator;
import com.tkr.engine.PipelineOrchestrator;
import com.tkr.types.WireTypes;
import com.tkr.types.WireTypes.*;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

/** CLI entry point for tkr-allocation-hub JVM tools. */
public final class TkrCtl {

    private TkrCtl() {}

    public static void main(String[] args) {
        if (args.length == 0) {
            printUsage();
            System.exit(1);
        }
        int code = switch (args[0]) {
            case "allocate" -> cmdAllocate(args);
            case "pipeline" -> cmdPipeline(args);
            case "router" -> cmdRouter(args);
            case "merge" -> cmdMerge(args);
            case "stats" -> cmdStats();
            default -> { printUsage(); yield 1; }
        };
        System.exit(code);
    }

    private static void printUsage() {
        System.err.println("Usage: tkrctl <command> [args]");
        System.err.println("Commands:");
        System.err.println("  allocate <total_qty> <weight1> <weight2> ...  Pro-rata allocation");
        System.err.println("  pipeline <file>                               Run allocation pipeline");
        System.err.println("  router <file>                                 Run router pipeline");
        System.err.println("  merge <file>                                  Merge session legs");
        System.err.println("  stats                                         Show pipeline stats");
    }

    private static int cmdAllocate(String[] args) {
        if (args.length < 3) {
            System.err.println("allocate requires total_qty and weights");
            return 1;
        }
        int totalQty = parseInt(args[1], -1);
        if (totalQty <= 0) return 1;
        List<ProRataAllocator.ProRataWeight> weights = new ArrayList<>();
        List<WireBatchRecord> records = new ArrayList<>();
        for (int i = 2; i < args.length; i++) {
            int w = parseInt(args[i], -1);
            if (w <= 0) return 1;
            ProRataAllocator.ProRataWeight pw = new ProRataAllocator.ProRataWeight();
            pw.accountId = 1000 + i;
            pw.weightBp = w * 100;
            weights.add(pw);
            WireBatchRecord rec = new WireBatchRecord();
            rec.recordId = i;
            rec.accountId = pw.accountId;
            rec.qtyMilli = pw.weightBp;
            records.add(rec);
        }
        ProRataAllocator allocator = new ProRataAllocator();
        ProRataAllocator.ProRataAllocatorResult result =
                allocator.allocateWithWeights(totalQty, weights, records);
        if (result.status != Status.OK) {
            System.err.println("allocation failed: " + WireTypes.statusToString(result.status));
            return 1;
        }
        System.out.println("Pro-rata allocation of " + totalQty + " units:");
        for (ProRataAllocator.ProRataSlice slice : result.slices) {
            System.out.println("  account " + slice.accountId + ": " + slice.qtyMilli);
        }
        System.out.println("Total allocated: " + result.totalAllocatedMilli);
        return 0;
    }

    private static int cmdPipeline(String[] args) {
        if (args.length < 2) return 1;
        byte[] data = readFile(args[1]);
        if (data == null) return 1;
        Status st = PipelineOrchestrator.runAllocationPipeline(data);
        System.out.println("pipeline: " + WireTypes.statusToString(st));
        return st == Status.OK ? 0 : 1;
    }

    private static int cmdRouter(String[] args) {
        if (args.length < 2) return 1;
        byte[] data = readFile(args[1]);
        if (data == null) return 1;
        Status st = PipelineOrchestrator.runRouterPipeline(data);
        System.out.println("router: " + WireTypes.statusToString(st));
        return st == Status.OK ? 0 : 1;
    }

    private static int cmdMerge(String[] args) {
        if (args.length < 2) return 1;
        byte[] data = readFile(args[1]);
        if (data == null) return 1;
        Status st = PipelineOrchestrator.mergeSessionLegs(data);
        System.out.println("merge: " + WireTypes.statusToString(st));
        return st == Status.OK ? 0 : 1;
    }

    private static int cmdStats() {
        PipelineOrchestrator.PipelineStats stats = PipelineOrchestrator.lastPipelineStats();
        System.out.println("batches_processed=" + stats.batchesProcessed);
        System.out.println("envelopes_sealed=" + stats.envelopesSealed);
        System.out.println("sessions_merged=" + stats.sessionsMerged);
        System.out.println("desk_calls=" + stats.deskCalls);
        return 0;
    }

    private static byte[] readFile(String path) {
        try {
            return Files.readAllBytes(Path.of(path));
        } catch (IOException e) {
            System.err.println("cannot read " + path + ": " + e.getMessage());
            return null;
        }
    }

    private static int parseInt(String s, int def) {
        try { return Integer.parseInt(s); } catch (NumberFormatException e) { return def; }
    }
}

File: legacy-cpp/tools/tkrctl.cc
#include "tkr/desk/allocation_solver.h"
#include "tkr/desk/pro_rata_allocator.h"
#include "tkr/engine/pipeline_orchestrator.h"
#include "tkr/types.h"
#include "tkr/util/bounds.h"
#include "tkr/wire/batch_wire_codec.h"

#include <cstdlib>
#include <iostream>
#include <string>
#include <vector>

namespace {

void PrintUsage(const char* prog) {
  std::cerr << "Usage: " << prog << " <command> [args]\n"
            << "Commands:\n"
            << "  allocate <total_qty> <weight1> <weight2> ...  Pro-rata allocation\n"
            << "  pipeline <file>                               Run allocation pipeline\n"
            << "  stats                                         Show pipeline stats\n";
}

int CmdAllocate(int argc, char** argv) {
  if (argc < 4) {
    std::cerr << "allocate requires total_qty and at least one weight\n";
    return 1;
  }

  std::int64_t total_qty = 0;
  if (!tkr::util::ParseAsciiInt(argv[2], &total_qty) || total_qty <= 0) {
    std::cerr << "invalid total_qty\n";
    return 1;
  }

  std::vector<tkr::desk::ProRataWeight> weights;
  std::vector<tkr::WireBatchRecord> records;

  for (int i = 3; i < argc; ++i) {
    std::int64_t weight = 0;
    if (!tkr::util::ParseAsciiInt(argv[i], &weight) || weight <= 0) {
      std::cerr << "invalid weight: " << argv[i] << "\n";
      return 1;
    }

    tkr::desk::ProRataWeight w{};
    w.account_id = static_cast<std::uint32_t>(1000 + i);
    w.weight_bp = static_cast<std::uint32_t>(weight * 100);
    weights.push_back(w);

    tkr::WireBatchRecord rec{};
    rec.record_id = static_cast<std::uint32_t>(i);
    rec.account_id = w.account_id;
    rec.qty_milli = w.weight_bp;
    records.push_back(rec);
  }

  tkr::desk::ProRataAllocator allocator(
      tkr::desk::ProRataAllocatorConfig{true});
  tkr::desk::ProRataAllocatorResult result = allocator.AllocateWithWeights(
      static_cast<std::uint32_t>(total_qty), weights, records);

  if (result.status != tkr::Status::kOk) {
    std::cerr << "allocation failed: " << tkr::StatusToString(result.status)
              << "\n";
    return 1;
  }

  std::cout << "Pro-rata allocation of " << total_qty << " units:\n";
  for (const tkr::desk::ProRataSlice& slice : result.slices) {
    std::cout << "  account " << slice.account_id << ": " << slice.qty_milli
              << "\n";
  }
  std::cout << "Total allocated: " << result.total_allocated_milli << "\n";
  return 0;
}

int CmdPipeline(const char* path) {
  FILE* fp = std::fopen(path, "rb");
  if (fp == nullptr) {
    std::cerr << "cannot open " << path << "\n";
    return 1;
  }

  std::fseek(fp, 0, SEEK_END);
  const long size = std::ftell(fp);
  std::fseek(fp, 0, SEEK_SET);

  if (size <= 0) {
    std::cerr << "empty file\n";
    std::fclose(fp);
    return 1;
  }

  std::vector<std::uint8_t> data(static_cast<std::size_t>(size));
  const std::size_t read =
      std::fread(data.data(), 1, data.size(), fp);
  std::fclose(fp);

  const tkr::Status st =
      tkr::engine::RunAllocationPipeline(data.data(), read);
  std::cout << "pipeline status: " << tkr::StatusToString(st) << "\n";
  return st == tkr::Status::kOk ? 0 : 1;
}

int CmdStats() {
  const tkr::engine::PipelineStats& stats =
      tkr::engine::LastPipelineStats();
  std::cout << "batches_processed=" << stats.batches_processed << "\n"
            << "envelopes_sealed=" << stats.envelopes_sealed << "\n"
            << "sessions_merged=" << stats.sessions_merged << "\n"
            << "desk_calls=" << stats.desk_calls << "\n";
  return 0;
}

}  // namespace

int main(int argc, char** argv) {
  if (argc < 2) {
    PrintUsage(argv[0]);
    return 1;
  }

  const std::string cmd = argv[1];
  if (cmd == "allocate") {
    return CmdAllocate(argc, argv);
  }
  if (cmd == "pipeline") {
    if (argc < 3) {
      std::cerr << "pipeline requires file path\n";
      return 1;
    }
    return CmdPipeline(argv[2]);
  }
  if (cmd == "stats") {
    return CmdStats();
  }

  PrintUsage(argv[0]);
  return 1;
}
